File: server/utility/createlab.pl
#!/usr/bin/perl -w
# Copyright (C) 2019-2024 Jan Frode Jæger <jan.frode.jaeger@ntnu.no>, NTNU, Trondheim, Norway
#
# This file is part of AURORA, a system to store and manage science data.
#
# AURORA is free software: you can redistribute it and/or modify it under 
# the terms of the GNU General Public License as published by the Free 
# Software Foundation, either version 3 of the License, or (at your option) 
# any later version.
#
# AURORA is distributed in the hope that it will be useful, but WITHOUT ANY 
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS 
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with 
# AURORA. If not, see <https://www.gnu.org/licenses/>. 
#
# Script to create a laboratory and its roles and permissions
#
use strict;
use JSON;
use Term::ReadKey;
use Data::Dumper;

my $CMD="./restcmd.pl";

my %opts;

print "AuthType [AuroraID]: ";
my $authtype=<STDIN>;
$authtype=~s/[\r\n]//g;
if ($authtype eq "") { $authtype="AuroraID"; }

print "Email-address: ";
my $email=<STDIN>;
$email=~s/[\r\n]//g;

print "AuthStr (not visible, for AuroraID - just the pw): ";
ReadMode('noecho'); 
my $authstr=ReadLine (0);
print "\n";
$authstr=~s/[\r\n]//g;
ReadMode('normal');

# put email and authstr together of AuroraID auth type
if ($authtype eq "AuroraID") { $authstr=$email.",".$authstr; }

print "Hostname: ";
my $host=<STDIN>;
$host=~s/[\r\n]//g;

print "Port [9393]: ";
my $port=<STDIN>;
$port=~s/[\r\n]//g;
$port=($port ne "" ? $port : 9393);

print "Private key: ";
my $key=<STDIN>;
$key=~s/[\r\n]//g;

print "Public key: ";
my $pkey=<STDIN>;
$pkey=~s/[\r\n]//g;

print "CA: ";
my $ca=<STDIN>;
$ca=~s/[\r\n]//g;

print "Parent Entity ID for Laboratory (Labs-group): ";
my $parent=<STDIN>;
$parent=~s/[\r\n]//g;

print "Laboratory-name (NTNU-FAC-DEP MyLabName): ";
my $name=<STDIN>;
$name=~s/[\r\n]//g;

# add to hash
$opts{authtype}=$authtype;
$opts{authstr}=$authstr;
$opts{host}=$host;
$opts{port}=$port;
$opts{key}=$key;
$opts{pkey}=$pkey;
$opts{ca}=$ca;
$opts{parent}=$parent;
$opts{name}=$name;

# we have all the information we need to create, but first test credentials
my $r;
#$r=execute (\%opts,"ping");
# ok, lets perform the operations needed to create the lab group
print "Creating laboratory group $name\n";
$r=execute (\%opts,"createGroup {\"name\":\"$name\",\"parent\":$parent}");
# get the created id
my $grid=$r->{id} || 0;
# create roles sub-group
print "Creating the roles sub-group\n";
$r=execute (\%opts,"createGroup {\"name\":\"roles\",\"parent\":$grid}");
my $rid=$r->{id} || 0;
# create permission/roles group
print "Creating the role group ${name}_user\n";
$r=execute (\%opts,"createGroup {\"name\":\"${name}_user\",\"parent\":$rid}");
my $usergr=$r->{id} || 0;
# set permissions on the role groups on the research group
print "Setting group permissions for ${name}_user on ${name}\n";
$r=execute (\%opts,"setGroupPerm {\"id\":$grid,\"user\":$usergr,\"grant\":[\"COMPUTER_READ\"]}");

print "Finished!...\n";

sub execute {
   my $opts=shift;
   my $command=shift;

   my $res=qx($CMD -t $opts->{authtype} -s $opts->{authstr} -h $opts->{host} -o $opts->{port} -k $opts->{key} -p $opts->{pkey} -c $opts->{ca} '$command');

   if ($? != 0) {
      print "   ERROR! Unable to execute script command: $command ($res). Aborting...\n";
      exit(1);
   }
   # add utf8-flag
   utf8::encode($res);
   # decode res
   my $h=decode($res);
   if (!defined $h) {
      exit(1);
   }
   # check success
   if ($h->{err} != 0) {
      print "   ERROR! Unable to call this REST-method: ".$h->{errstr}."\n";
   }
   return $h;
}

sub decode {
   my $content=shift || "";

   my $err;
   my $h;
   local $@;
   eval { my $a=JSON->new(); $h=$a->decode ($content); };
   $@ =~ /nefarious/ and $err=$@;

   if (!defined $err) {
      # decoding was a success - return structure
      return $h;
   } else {
      # an error occured
      print "Unable to decode JSON: $err\n";
      return undef;
   }
}

File: client/lua-scripts/create_lab.lua
--
-- Copyright(C) 2019-2024 Jan Frode Jæger, NTNU, Trondheim, Norway
--
-- This file is part of AURORA, a system to store and manage science data.
--
-- AURORA is free software: you can redistribute it and/or modify it under 
-- the terms of the GNU General Public License as published by the Free 
-- Software Foundation, either version 3 of the License, or (at your option) 
-- any later version.
--
-- AURORA is distributed in the hope that it will be useful, but WITHOUT ANY 
-- WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS 
-- FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along with 
-- AURORA. If not, see <https://www.gnu.org/licenses/>. 
--
-- create_lab: script to create a NTNU lab.
--
parent=readstr("What is the parent ID of the lab?\n");
name=readstr("What is the lab name?\n");
print ("Parent of lab: " .. parent .. "\n");
print ("Lab name: " .. name .. "\n");
ok = readstr("Do you want to create lab? [Y/N]\n","Y");
if (string.upper(ok) == "Y") then
   print ("Creating lab " .. name .. " group\n");
   params={};
   params.parent=parent;
   params.name=name;
   lresult=aurora("createGroup",params);
   if (lresult.err == 0) then
      -- create roles group under the lab      
      print ("   Successfully created lab " .. name .."\n");
      print ("Creating roles group on lab...\n");
      labid = lresult.id;
      rparams={};
      rparams.parent=labid;
      rparams.name="roles";
      rresult=aurora("createGroup",rparams);
      if (rresult.err == 0) then
         -- create user roles group
         print ("   Successfully created roles group under lab" .. "\n");
         print ("Creating user group under roles...\n");
         urparams={};
         urparams.parent=rresult.id;
         urparams.name=name .. "_user";
         urres=aurora("createGroup",urparams);
         if (urres.err == 0) then
            -- create admin roles group
            print ("   Successfully created user roles group " .. urparams.name .. " for lab " .. name .. "\n");
            print ("Creating admin group under roles...\n");
            arparams={};
            arparams.parent=rresult.id;
            arparams.name=name .. "_admin";
            arres=aurora("createGroup",arparams);
            if (arres.err == 0) then
               -- set perms
               print ("   Successfully created admin roles group " .. arparams.name .. " for lab " .. name .. "\n");
               -- set perms for users on lab
               print ("Setting permissions for users on lab...\n");
               uperms={};
               uperms.user=urres.id;
               uperms.id=labid;
               uperms.grant={"COMPUTER_READ"};
               pures=aurora("setGroupPerm",uperms);
               -- set perms for admins on lab
               print ("Setting permissions for admins on lab...\n");
               aperms={};
               aperms.user=arres.id;
               aperms.id=labid;
               aperms.grant={"COMPUTER_CREATE","COMPUTER_CHANGE","COMPUTER_MOVE","COMPUTER_READ"};
               pares=aurora("setGroupPerm",aperms);
               -- set perms for admins on roles group
               print ("Setting permissions for admins on roles group...\n");
               rperms={};
               rperms.user=arres.id;
               rperms.id=rresult.id;
               rperms.grant={"GROUP_MEMBER_ADD"};
               rperms.deny={"COMPUTER_CREATE"};
               prres=aurora("setGroupPerm",rperms);
               if ((pures.err == 0) and (pares.err == 0) and (prres.err == 0)) then
                  print ("   Successfully set all permissions for lab...\n");
                  print ("   Successfully completed creating lab " .. name .."\n");
               end
            end
         end
      end
   end
end
